### ISSUE TITLE
salt-master fails to communicate with salt-minion after upgrade to 2015.8.0 (until manual salt-minion restart)

### ISSUE THREAD

**Issue by yermulnik** [saltstack/salt#27267]
While upgrading my minions to 2015.8.0 via SaltStack I had encountered a problem that I have to manually restart minions because master fails to communicate with a newly upgraded minion:
```yaml
jail_centos:
----------
          ID: salt-minion
    Function: pkgrepo.managed
        Name: saltstack
      Result: True
     Comment: Configured package repo 'saltstack'
     Started: 10:05:07.325135
    Duration: 14.847 ms
     Changes:
              ----------
              repo:
                  saltstack
----------
          ID: salt-minion
    Function: pkg.latest
      Result: True
     Comment: The following packages were successfully installed/upgraded: salt-minion
     Started: 10:05:07.340079
    Duration: 17572.675 ms
     Changes:
              ----------
              python-futures:
                  ----------
                  new:
                      3.0.3-1.el7
                  old:
              python-tornado:
                  ----------
                  new:
                      4.2.1-1.el7
                  old:
              salt:
                  ----------
                  new:
                      2015.8.0-2.el7
                  old:
                      2015.5.5-1.el7
              salt-minion:
                  ----------
                  new:
                      2015.8.0-2.el7
                  old:
                      2015.5.5-1.el7
----------
          ID: salt-minion
    Function: service.running
      Result: False
     Comment: An exception occurred in this state: Traceback (most recent call last):
                File "/usr/lib/python2.7/site-packages/salt/state.py", line 1561, in call
                  # If the state function accepts an 'env' keyword argument, it
                File "/usr/lib/python2.7/site-packages/salt/states/service.py", line 504, in mod_watch
                  if __salt__['service.status'](name, sig):
                File "/usr/lib/python2.7/site-packages/salt/modules/systemd.py", line 521, in status
                  if _untracked_custom_unit_found(name) or _unit_file_changed(name):
                File "/usr/lib/python2.7/site-packages/salt/modules/systemd.py", line 135, in _untracked_custom_unit_found
                  return name not in get_all() and os.access(unit_path, os.R_OK)
                File "/usr/lib/python2.7/site-packages/salt/modules/systemd.py", line 338, in get_all
                  + list(_get_all_legacy_init_scripts().keys())))
                File "/usr/lib/python2.7/site-packages/salt/modules/systemd.py", line 77, in _get_all_units
                  python_shell=True
                File "/usr/lib/python2.7/site-packages/salt/modules/cmdmod.py", line 1186, in run_stdout
                  use_vt=use_vt)
                File "/usr/lib/python2.7/site-packages/salt/modules/cmdmod.py", line 454, in _run
                  out = salt.utils.to_str(out).rstrip()
              AttributeError: 'module' object has no attribute 'to_str'
     Started: 10:05:25.831995
    Duration: 9.401 ms
     Changes:

Summary for jail_centos
------------
Succeeded: 2 (changed=2)
Failed:    1
------------
Total states run:     3
Total run time:  18.480 s
```
It's not very comfortable when one has a lot of minions =(

**Comment by jfindlay** [comment]
@yermulnik, thanks for the report.  What version are you running on your master?  We [recommend](https://docs.saltstack.com/en/latest/faq.html#can-i-run-different-versions-of-salt-on-my-master-and-minion) upgrading the master before the minions for situations like this to narrow down the variability of incompatible changes, although it is possible that we missed something.

**Comment by cachedout** [comment]
How did you install the minion? Was this a package supplied from `repo.saltstack.com`?

**Comment by yermulnik** [comment]
@jfindlay it's 2015.8.0 (Beryllium) on FreeBSD 10.1 amd64. Master was upgraded on September 15. I also tried to restart minion via `service.restart` execution module (CLI), but got the same error. Should I use exactly `cmd.wait` (`echo ... | at now ...`)?
@cachedout I had encountered this issue on FreeBSD (master and minion were both installed from FreeBSD Ports Collection) and on CentOS with minion installed from `repo.saltstack.com` via `pkgrepo.managed` and `pkg.latest` (as you can see from my first post). Can't check it on Ubuntu as there're still no actual packages for this OS.

**Comment by yermulnik** [comment]
FYI: today I had upgraded minions to 2015.8.0 on my Ubuntu servers (trusty) from `repo.saltstack.com` and there were no errors. But I had to use execution module `pkg.install` instead of state module `pkg.latest` as `pkg.latest` couldn't see the new version available (see https://github.com/saltstack/salt/issues/27159#issuecomment-144006634).

**Comment by jagguli** [comment]
im getting the same error with salt-ssh

  File "/tmp/.root_022bd0_salt/salt-call", line 4, in <module>
            salt_call()
          File "/tmp/.root_022bd0_salt/salt/scripts.py", line 330, in salt_call
            client = salt.cli.call.SaltCall()
          File "/tmp/.root_022bd0_salt/salt/utils/parsers.py", line 128, in __init__
            optparse.OptionParser.__init__(self, *args, **kwargs)
          File "/usr/lib/python2.7/optparse.py", line 1219, in __init__
            add_help=add_help_option)
          File "/tmp/.root_022bd0_salt/salt/utils/parsers.py", line 204, in _populate_option_list
            mixin_setup_func(self)
          File "/tmp/.root_022bd0_salt/salt/utils/parsers.py", line 1029, in _mixin_setup
            config.minion_config(None)
          File "/tmp/.root_022bd0_salt/salt/config/__init__.py", line 1555, in minion_config
            opts = apply_minion_config(overrides, defaults, cache_minion_id=cache_minion_id)
          File "/tmp/.root_022bd0_salt/salt/config/__init__.py", line 2700, in apply_minion_config
            cache_minion_id=cache_minion_id)
          File "/tmp/.root_022bd0_salt/salt/config/__init__.py", line 2669, in get_id
            newid = salt.utils.network.generate_minion_id()
          File "/tmp/.root_022bd0_salt/salt/utils/network.py", line 284, in generate_minion_id
            for addr in salt.utils.network.ip_addrs():
          File "/tmp/.root_022bd0_salt/salt/utils/network.py", line 993, in ip_addrs
            return _ip_addrs(interface, include_loopback, interface_data, 'inet')
          File "/tmp/.root_022bd0_salt/salt/utils/network.py", line 968, in _ip_addrs
            else interfaces()
          File "/tmp/.root_022bd0_salt/salt/utils/network.py", line 778, in interfaces
            return linux_interfaces()
          File "/tmp/.root_022bd0_salt/salt/utils/network.py", line 664, in linux_interfaces
            salt.utils.to_str(cmd1),
        AttributeError: 'module' object has no attribute 'to_str'

**Comment by zmalone** [comment]
This problem also seems to appear on Ubuntu 14.04LTS with the repo.saltstack.com repo on a fresh install of both a master and minion.

**Comment by cachedout** [comment]
@zmalone That is...odd. I've noticed Python have problems with detecting changes in package-level functions that get added in removed when byte-caching files, but that wouldn't apply on a new installation. If you pop open the salt.utils init file, do you see the function in question?

**Comment by lburg** [comment]
Hi, I get the same error running a highstate with a `service.running` state. I'm running salt-minion 2015.8.1 installed using the bootstrap-salt.sh script, and salt-master 2015.8.1 installed through `repo.saltstack.com`. Both the minion and the master are running on Debian. If I open `salt/utils/__init__.py`, I do see the `to_str` function at line 2776. Also, running the highstate a second time yields no error and completes succesfully. Here is an extract of my logs:

```
Nov 12 09:45:33 <user> bash[145]: [INFO    ] Running state [dbus] at time 09:45:33.102708
Nov 12 09:45:33 <user> bash[145]: [INFO    ] Executing state service.running for dbus
Nov 12 09:45:33 <user> bash[145]: [INFO    ] Executing command 'systemctl --all --full --no-legend --no-pager list-units | col -b' in directory '/root'
Nov 12 09:45:33 <user> bash[145]: [ERROR   ] An exception occurred in this state: Traceback (most recent call last):
Nov 12 09:45:33 <user> bash[145]: File "/usr/lib/python2.7/dist-packages/salt/state.py", line 1560, in call
Nov 12 09:45:33 <user> bash[145]: #
Nov 12 09:45:33 <user> bash[145]: File "/usr/lib/python2.7/dist-packages/salt/states/service.py", line 290, in running
Nov 12 09:45:33 <user> bash[145]: if not _available(name, ret):
Nov 12 09:45:33 <user> bash[145]: File "/usr/lib/python2.7/dist-packages/salt/states/service.py", line 248, in _available
Nov 12 09:45:33 <user> bash[145]: avail = __salt__['service.available'](name)
Nov 12 09:45:33 <user> bash[145]: File "/usr/lib/python2.7/dist-packages/salt/modules/systemd.py", line 355, in available
Nov 12 09:45:33 <user> bash[145]: units = get_all()
Nov 12 09:45:33 <user> bash[145]: File "/usr/lib/python2.7/dist-packages/salt/modules/systemd.py", line 338, in get_all
Nov 12 09:45:33 <user> bash[145]: + list(_get_all_legacy_init_scripts().keys())))
Nov 12 09:45:33 <user> bash[145]: File "/usr/lib/python2.7/dist-packages/salt/modules/systemd.py", line 77, in _get_all_units
Nov 12 09:45:33 <user> bash[145]: python_shell=True
Nov 12 09:45:33 <user> bash[145]: File "/usr/lib/python2.7/dist-packages/salt/modules/cmdmod.py", line 1186, in run_stdout
Nov 12 09:45:33 <user> bash[145]: use_vt=use_vt)
Nov 12 09:45:33 <user> bash[145]: File "/usr/lib/python2.7/dist-packages/salt/modules/cmdmod.py", line 454, in _run
Nov 12 09:45:33 <user> bash[145]: out = salt.utils.to_str(out).rstrip()
Nov 12 09:45:33 <user> bash[145]: AttributeError: 'module' object has no attribute 'to_str'
Nov 12 09:45:33 <user> bash[145]: [INFO    ] Completed state [dbus] at time 09:45:33.115166
Nov 12 09:45:33 <user> bash[145]: [ERROR   ] An un-handled exception was caught by salt's global exception handler:
Nov 12 09:45:33 <user> bash[145]: TypeError: get_colors() takes at most 1 argument (2 given)
Nov 12 09:45:33 <user> bash[145]: Traceback (most recent call last):
Nov 12 09:45:33 <user> bash[145]: File "/usr/bin/salt-call", line 11, in <module>
Nov 12 09:45:33 <user> bash[145]: salt_call()
Nov 12 09:45:33 <user> bash[145]: File "/usr/lib/python2.7/dist-packages/salt/scripts.py", line 227, in salt_call
Nov 12 09:45:33 <user> bash[145]: return
Nov 12 09:45:33 <user> bash[145]: File "/usr/lib/python2.7/dist-packages/salt/cli/call.py", line 69, in run
Nov 12 09:45:33 <user> bash[145]: File "/usr/lib/python2.7/dist-packages/salt/cli/caller.py", line 246, in run
Nov 12 09:45:33 <user> bash[145]: pass
Nov 12 09:45:33 <user> bash[145]: File "/usr/lib/python2.7/dist-packages/salt/output/__init__.py", line 94, in display_output
Nov 12 09:45:33 <user> bash[145]: fdata = display_data
Nov 12 09:45:33 <user> bash[145]: File "/usr/lib/python2.7/dist-packages/salt/output/__init__.py", line 46, in try_printout
Nov 12 09:45:33 <user> bash[145]: return get_printout('raw', opts)(data).rstrip()
Nov 12 09:45:33 <user> bash[145]: File "/usr/lib/python2.7/dist-packages/salt/output/highstate.py", line 85, in output
Nov 12 09:45:33 <user> bash[145]: return _format_host(host, hostdata)[0]
Nov 12 09:45:33 <user> bash[145]: File "/usr/lib/python2.7/dist-packages/salt/output/highstate.py", line 91, in _format_host
Nov 12 09:45:33 <user> bash[145]: __opts__.get('color_theme'))
Nov 12 09:45:33 <user> bash[145]: TypeError: get_colors() takes at most 1 argument (2 given)
Nov 12 09:45:33 <user> bash[145]: Traceback (most recent call last):
Nov 12 09:45:33 <user> bash[145]: File "/usr/bin/salt-call", line 11, in <module>
Nov 12 09:45:33 <user> bash[145]: salt_call()
Nov 12 09:45:33 <user> bash[145]: File "/usr/lib/python2.7/dist-packages/salt/scripts.py", line 227, in salt_call
Nov 12 09:45:33 <user> bash[145]: return
Nov 12 09:45:33 <user> bash[145]: File "/usr/lib/python2.7/dist-packages/salt/cli/call.py", line 69, in run
Nov 12 09:45:33 <user> bash[145]: File "/usr/lib/python2.7/dist-packages/salt/cli/caller.py", line 246, in run
Nov 12 09:45:33 <user> bash[145]: pass
Nov 12 09:45:33 <user> bash[145]: File "/usr/lib/python2.7/dist-packages/salt/output/__init__.py", line 94, in display_output
Nov 12 09:45:33 <user> bash[145]: fdata = display_data
Nov 12 09:45:33 <user> bash[145]: File "/usr/lib/python2.7/dist-packages/salt/output/__init__.py", line 46, in try_printout
Nov 12 09:45:33 <user> bash[145]: return get_printout('raw', opts)(data).rstrip()
Nov 12 09:45:33 <user> bash[145]: File "/usr/lib/python2.7/dist-packages/salt/output/highstate.py", line 85, in output
Nov 12 09:45:33 <user> bash[145]: return _format_host(host, hostdata)[0]
Nov 12 09:45:33 <user> bash[145]: File "/usr/lib/python2.7/dist-packages/salt/output/highstate.py", line 91, in _format_host
Nov 12 09:45:33 <user> bash[145]: __opts__.get('color_theme'))
Nov 12 09:45:33 <user> bash[145]: TypeError: get_colors() takes at most 1 argument (2 given)
```

I guess the second error, `TypeError: get_colors() takes at most 1 argument (2 given)`, is a result of the first one but I included it in case it's of any help.

**Comment by cachedout** [comment]
@lburg Was that installation an upgrade or a fresh one?

**Comment by basepi** [comment]
@lburg Can you reply to @cachedout's latest comment please?

**Comment by lburg** [comment]
Sorry about the late reply, I shouldn't have marked the notification as read!

Our minion is installed using a Debian preseed so it's always a fresh install, including the OS. However since then we moved from bootstrap-salt.sh and install the minion using our own debian repository, we don't have the issue anymore.

We changed many things since then and I cannot manage to reproduce the error. I've checked out an old commit from our project around November 12 and ran a highstate but to no avail.

Here is the state that initially produced the error:

```yaml
dbus:
  service.running
```
